### ISSUE TITLE
The autograd.softmax has some problems

### ISSUE THREAD

**Issue by chrishkchris** [apache/singa#578]
The autograd.softmax may have some problem, where I found it when I took part in the review of PR #572.

In examples/autograd/mlp.py (multilayer perception), the result is:

```
ubuntu@ip-172-31-26-47:~/singa/examples/autograd$ python3 mlp.py
train_data_shape: (400, 2)
train_label_shape: (400, 2)
training loss =  0.6908062
training loss =  0.5960194
training loss =  0.57797414
training loss =  0.55334115
training loss =  0.48568404
training loss =  0.38458923
training loss =  0.30776194
training loss =  0.24188559
training loss =  0.18657134
training loss =  0.15864176
training loss =  0.13929243
```

However, if I use softmax + cross_entropy instead of softmax_cross_entropy, there is such error:

```
ubuntu@ip-172-31-26-47:~/singa/examples/autograd$ python3 mlp.py
train_data_shape: (400, 2)
train_label_shape: (400, 2)
training loss =  6.682101
WARNING: Logging before InitGoogleLogging() is written to STDERR
F0113 09:20:05.180658 12032 tensor_math_cpp.h:357] Check failed: a > 0.f (-nan vs. 0)
*** Check failure stack trace: ***
Aborted (core dumped)
```

In the review of PR #572, I did not suspect SoftMax because I compared the 1-D result with Pytorch. However, now when I run the result with 2-D input, the backpropagation cannot with parameter axis = 1.

My test codes and results for softmax are in:
https://gist.github.com/chrishkchris/1bce55260b5e771ce974940a855292e2

I will need to further see how to debug.

**Comment by chrishkchris** [comment]
This error will be resolved using PR #588

**Comment by chrishkchris** [comment]
It does not hangs, it is just too slow, much much slower than the previous version.

**Comment by chrishkchris** [comment]
addressed by PR #588
